Mathlib/Algebra/Category/ModuleCat/Sheaf/Generators.lean

```diff
@@ -73,4 +88,39 @@ def equivOfIso (e : M ≅ N) :
 
 end GeneratingSections
 
+variable [∀ (X : C), HasWeakSheafify (J.over X) AddCommGrp.{u}]
+  [∀ (X : C), (J.over X).WEqualsLocallyBijective AddCommGrp.{u}]
+  [∀ (X : C), (J.over X).HasSheafCompose (forget₂ RingCat AddCommGrp.{u})]
+
+/-- The data of generating sections of the restriction of a sheaf of modules
+over a covering of the terminal object. -/
+structure LocalGeneratorsData where
+  /-- the index type of the covering -/
+  I : Type u'
+  /-- a family of objects which cover the terminal object -/
+  X : I → C
+  coversTop : J.CoversTop X
+  /-- the data of sections of `M` over `X i` which generate `M.over (X i)` -/
+  generators (i : I) : (M.over (X i)).GeneratingSections
+
+/-- A sheaf of modules is of finite type if locally, it is generated by finitely
+many sections. -/
+class IsFiniteType : Prop where
+  exists_localGeneratorsData :
+    ∃ (σ : M.LocalGeneratorsData), ∀ (i : σ.I), Finite (σ.generators i).I
```

Comment:
If I understood it right, then the following thing would be `IsFiniteType`: C is the category associated to some infinite discrete space like Nat, with sheaf of modules Int=Z, and attached to n you have the Z-module Z^n. Then you can cover with Nat (i.e. sigma.X=id) and you have something which I would instinctively say was "locally finite type", but *not* finite type. Is this notation standard or are we just making stuff up for mathlib here? If this is our own invention then maybe `IsLocallyFiniteType` is a better name? What do you think?

Reply:
Your example is indeed of finite type. This is standard terminology: "finite type" means locally of finite type, see https://stacks.math.columbia.edu/tag/01B5

Here, I think that there is a parallel with relative properties of morphisms of schemes: we should think that to be a finite type sheaf of modules is a relative property "over" the site. For example, in the language of schemes, if you construct the scheme `S` which is a coproduct of countably many copies of the spectrum of the complex numbers (the underlying topological space is the discrete set `Nat`), then the scheme `X` which is a disjoint union of affine spaces of dimension `n` for all `n : Nat` over `ℂ`, then you get a morphism `X ⟶ S` in a way that the fibre over `n : Nat` (identified to `S`) is the `n`-dimensional affine space over `ℂ`. Yet, `X` is a finite type `S`-scheme, not just locally of finite type! (As a scheme over `ℂ` in an obvious way, `X` is no longer of finite type, but it is locally of finite type...)